### ISSUE TITLE
Be more specific in buffers creation

### ISSUE THREAD

**Issue by tomaka** [tomaka/glium#882]
I feel that the `dynamic()` method creating persistent mapped buffers is too high-level/confusing.

Instead we should add `persistent_if_supported`, `persistent`, and `persistent_or_dynamic`.

**Comment by tomaka** [comment]
So thing thing needs a decision.

There are 5 relevant ways to create/manager a buffer:

 - A regular old-fashioned buffer with `glBufferData(..., GL_STATIC_DRAW)`.
 - A regular old-fashioned buffer with `glBufferData(..., GL_DYNAMIC_DRAW)`, slightly more optimized for frequent updates.
 - With `glBufferStorage(..., GL_DYNAMIC_STORAGE_BIT)` with or without `GL_CLIENT_STORAGE_BIT`.
 - The new way, with `glBufferStorage()` without any flag.
 - The new way, with `glBufferStorage(..., GL_MAP_PERSISTENT)`.

In the first three situations, the buffer can be freely reallocated by the driver. This means that there's absolutely no problem if the user writes repeaditely to the same part of the buffer. Each write can in reality allocate a new buffer. I call this the "old-fashioned way" because the OpenGL driver does a lot of magic.

The fourth way (`glBufferStorage` without flags) creates a buffer that won't be reallocated. Calling any function except `glCopyBufferSubData` and `glClearBufferSubData` is forbidden. Glium allows one to bypass these restrictions by creating temporary buffers and copying them with `glCopyBufferSubData`. This solution is faster than the first three if the buffer doesn't change its content, and slower if you change it.

The last way creates a buffer that won't be reallocated and that can be manually written and synchronized. This is by far the fastest way if you just upload data once per frame. However it has the problem that you can't write to a part of the buffer that is still in use by the GPU. Writing repeatedly to the same buffer will be **very** slow. It also handles fences manually, so you must try to write big chunks at once instead of writing individual elements.

To sum things up:
 - If you never touch the content of your buffer again, the fastest ways are 4 then 1 and 3 without client_storage (1 and 3 should be equal).
 - If you change the content of your buffer once per frame, or if you use the triple buffering technique, the fastest way is 5.
 - If you change the content of your buffer repeatedly (upload, draw, upload, draw, etc.) then 2 and 3 are the fastest way.

For moment, calling `Buffer::new()` uses 4 and falls back to 1 if it's not available. Calling `Buffer::dynamic()` uses 5 and falls back to 2 if it's not available.
Therefore the situation where you change the content of your buffer repeatedly you have no appropriate way.

**Comment by tomaka** [comment]
In addition to this, there are also options for the buffer's mapping. One can choose to have a buffer that can only be mapped for reading or writing, for better performances.

I suggest the following:

 - Calling `Buffer::dynamic()` uses 3 (with client_storage) and falls back to 2.
 - Calling `Buffer::new()` uses 3 (without client_storage) and falls back to 1.
 - Add `Buffer::persistent()` that uses 5 and falls back to 2, with an enum that lets you choose between `ReadOnly`, `WriteOnly`, and `ReadWrite`. Optimized for writing once per frame or for manually using triple buffering.
 - Add `Buffer::immutable()` that uses 4 and falls back to 1. Optimized for when you never change the content of your buffer.

**Comment by tomaka** [comment]
I suggest the following:

 - Calling `Buffer::new()` uses 3 (without client_storage) and falls back to 1. Always allow read + write. This is the dummy way for people who are lazy.
 - Calling `Buffer::dynamic()` uses 3 (with client_storage) and falls back to 2, with an enum that lets you choose between `ReadOnly`, `WriteOnly`, and `ReadWrite`.
 - Add `Buffer::persistent()` that uses 5 and falls back to 2, with the same enum as `dynamic()`.
 - Add `Buffer::immutable()` that uses 4 and falls back to 1. Optimized for when you never change the content of your buffer.
